### ISSUE TITLE
React Vite compatibility 

### ISSUE THREAD

**Issue by pdrvsky** [fullcalendar/fullcalendar-react#150]
Hello,

Recently we've been researching a possibility to move our project to Vite environment. Of course we failed at the first try just because of the same issue as discussed in the Vue integration section of your repository (i.e. https://github.com/fullcalendar/fullcalendar-vue/issues/152 and related). I've tried the solution provided there to import `@fullcalendar/core/vdom` just before anything else connected to `fullcalendar` packages but it didn't work very well for a React project. The error I got was:

`The <CalendarDataProvider2 /> component appears to have a render method, but doesn't extend React.Component. This is likely to cause errors. Change CalendarDataProvider2 to extend React.Component instead. `

and:

`Uncaught TypeError: Cannot set property 'props' of undefined`.

It definitely concerned the imported vdom not being compatible with Reacts vDOM. So I crawled through internal `fullcalendar` package files and I found `@fullcalendar/react/dist/vdom`. When replacing the import the project seems to work.

To sum up I would like to ask the project authors if the described solution is a proper way to run a React Fullcalendar project with a Vite build tool? If so, could the docs be updated (or should I prepare a PR to update them)?

**Comment by Maadtin** [comment]
Any news on this?

**Comment by abdellatifLabr** [comment]
Getting the same error, I hope this gets fixed as soon as possible.

**Comment by abdellatifLabr** [comment]
For everyone who is still getting this error, here is the fix
You should put your imports in this order including react vdom
```js
import "@fullcalendar/react/dist/vdom";
import FullCalendar from "@fullcalendar/react";
import dayGridPlugin from "@fullcalendar/daygrid";
```

**Comment by Maadtin** [comment]
@abdellatifLabr  Still getting the same error :S

**Comment by abdellatifLabr** [comment]
@Maadtin the solution that worked for me is the one i posted above

**Comment by jugglingcats** [comment]
I'm running vite too and @abdellatifLabr's sequence above worked for me (using `/react/dist/vdom`). For me using `/core/vdom` _did get rid of the first error_ about import being in the wrong order but I had the following cryptic warnings/errors...

```
Warning: CalendarRoot2 defines an invalid contextType. contextType should point to the Context object returned by React.createContext(). However, it is set to an object with keys {__c, __, Consumer, Provider}.
...
Uncaught Error: Expected ref to be a function, a string, an object returned by React.createRef(), or null.
...
The above error occurred in the <CalendarContent2> component
```
(for benefit of others coming here via google)

**Comment by benj56** [comment]
After spending a few hours on this, I now realize that @jugglingcats already pointed out the difference between the correct `import '@fullcalendar/react/dist/vdom';` and the incorrect `import '@fullcalendar/core/vdom';`. 

The latter imports from `core-preact`, loading preact `createRef` etc. instead of the React functions, leading to the cryptic errors.

**Comment by SnowingFox** [comment]
> For everyone who is still getting this error, here is the fix You should put your imports in this order including react vdom
> 
> ```js
> import "@fullcalendar/react/dist/vdom";
> import FullCalendar from "@fullcalendar/react";
> import dayGridPlugin from "@fullcalendar/daygrid";
> ```

it works well, but why should i to add this line?
```js
import "@fullcalendar/react/dist/vdom";
```

can u explain it? im curious about it

**Comment by SnowingFox** [comment]
> > can u explain it? im curious about it
> 
> See [fullcalendar/fullcalendar#6371](https://github.com/fullcalendar/fullcalendar/issues/6371). With Vite, in the dev build, the pre-bundled fullcalendar dependency you import does not have the order fullcalendar expects.
> 
> By importing that file manually from the fullcalendar dist folder, it will set the required globals before loading the pre-bundled fullcalendar bundle in your next import.

thanks, an interesting problem

**Comment by brittaniSavery** [comment]
> For everyone who is still getting this error, here is the fix You should put your imports in this order including react vdom
> 
> ```js
> import "@fullcalendar/react/dist/vdom";
> import FullCalendar from "@fullcalendar/react";
> import dayGridPlugin from "@fullcalendar/daygrid";
> ```

Not sure if this is a fullcalendar issue or an [astro ](astro.build) issue, but when I added the `import "@fullcalendar/react/dis/vdom";` line, it did remove the import ordering error. However, I'm now getting this error instead.

```
Cannot use import statement outside a module

C:\...\node_modules\react\dist\vdom.js:1
import * as react from 'react';
^^^^^^

Syntax Error: Cannot use import statement outside a module
         at wrapSafe (internal/modules/cjs/loader.js:979:16)
```

**Comment by dangdn721** [comment]
> For everyone who is still getting this error, here is the fix You should put your imports in this order including react vdom
> 
> ```js
> import "@fullcalendar/react/dist/vdom";
> import FullCalendar from "@fullcalendar/react";
> import dayGridPlugin from "@fullcalendar/daygrid";
> ```

Thanks you. It's work on mode development. But when I using vite build it's still happened. Any one help me ?

**Comment by dangdn721** [comment]
> > can u explain it? im curious about it
> 
> See [fullcalendar/fullcalendar#6371](https://github.com/fullcalendar/fullcalendar/issues/6371). With Vite, in the dev build, the pre-bundled fullcalendar dependency you import does not have the order fullcalendar expects.
> 
> By importing that file manually from the fullcalendar dist folder, it will set the required globals before loading the pre-bundled fullcalendar bundle in your next import.

Thank you for answer. But Do you know how to fix order import for fullcalendar in pre-bundled Vite ? I have isssue about it.

**Comment by Digital-Coder** [comment]
in vite `2.9.x` `import "@fullcalendar/react/dist/vdom";` solution has been working ,but in `vite@3.0.2` I am receiving that react is not defined using fullcallendar. Anyone figured it out how to solve it ?

**Comment by adrianfabjanski** [comment]
> Digital

Same issue here. `React is not defined` error occurring in automatic jsxRuntime on production build after upgrading vite.
![image](https://user-images.githubusercontent.com/68193533/181190197-bdfaa95c-9668-49a3-82b7-8467210705e7.png)

any plans on fixing it?

**Comment by tgiovanella87** [comment]
> For everyone who is still getting this error, here is the fix You should put your imports in this order including react vdom
> 
> ```js
> import "@fullcalendar/react/dist/vdom";
> import FullCalendar from "@fullcalendar/react";
> import dayGridPlugin from "@fullcalendar/daygrid";
> ```

It really Works! Thanks a lot!!!

**Comment by Srounsroun** [comment]
@adrianfabjanski  do you use @vitejs/plugin-react ?

```
import react from '@vitejs/plugin-react'
export default defineConfig({
    plugins: [react()]
})
```

**Comment by tgiovanella87** [comment]
Yes I am using React with Vite and my file starts like that:

`import React, { useRef } from "react";
import "@fullcalendar/react/dist/vdom";
import "@fullcalendar/core/locales/pt-br";
import FullCalendar from "@fullcalendar/react"; // must go before plugins
import dayGridPlugin from "@fullcalendar/daygrid"; // a plugin!
import listGridPlugin from "@fullcalendar/list";
import interactionPlugin from "@fullcalendar/interaction";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";`

And than, the Component is something like that

`<FullCalendar
                plugins={[dayGridPlugin, listGridPlugin, interactionPlugin]}
                initialView="dayGridMonth"
                height={500}
                locale="pt-br"
                headerToolbar={{
                  start: "listWeek,dayGridMonth,today,prev,next",
                  center: "title",
                  end: "prevYear,nextYear",
                }}
                dateClick={(e) => handleDateClick(e)}
                eventClick={(e) => handleEventClick(e)}
                nowIndicator={true}
                buttonText={{
                  today: "Hoje",
                  month: "Mês",
                  week: "Semana",
                  day: "Dia",
                  list: "Lista",
                }}
              />`

**Comment by arshaw** [comment]
Fixed in [v6.0.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v6.0.0)

The vdom import workaround is no longer necessary.

**Comment by lccmanuel** [comment]
you have right @arshaw in v6 no more is needed, just I remove de line 
`import "@fullcalendar/react/dist/vdom"`
and it worked again :)

**Comment by Karan-The-Coder** [comment]
DayTableView defines an invalid contextType. contextType should point to the Context object returned by React.createContext(). However, it is set to an object with keys {__c, __, Consumer, Provider}. 

I got this error when I use fullcalendar `components.`

**Comment by TapendraPNGTS** [comment]
[plugin:vite:import-analysis] Missing "./dist/vdom" specifier in "@fullcalendar/react" package
same problem
